Stats Improvements Step 2: add date bar to the top of the charts

anitaa1990 opened the pull request:
Fixes #1253 . This PR completes the second step of the Stats Improvements Project as defined in the master thread [here](https://github.com/woocommerce/woocommerce-android/issues/1199).

### Changes
This PR adds the date bar to both the old and the new stats UI.

### Behaviour
(LEFT: old stats .  RIGHT: v4 stats)
<img width="300" src="https://user-images.githubusercontent.com/22608780/62442378-5bc1e080-b775-11e9-9f2b-915529dba198.gif"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/62442429-8875f800-b775-11e9-9120-08d07fa41428.gif">

### Screenshots 
(LEFT: old stats .  RIGHT: v4 stats)
<img width="300" src="https://user-images.githubusercontent.com/22608780/62442456-9af03180-b775-11e9-8e8f-9caa27af494f.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/62442466-9fb4e580-b775-11e9-9719-a10700ee3d49.png">

<img width="300" src="https://user-images.githubusercontent.com/22608780/62442479-af342e80-b775-11e9-9d1b-9dfd121102c5.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/62442488-b2c7b580-b775-11e9-8f18-4cff2589e03a.png">

<img width="300" src="https://user-images.githubusercontent.com/22608780/62442502-bfe4a480-b775-11e9-93a7-6023447fe17f.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/62442507-c246fe80-b775-11e9-9fa1-947bafccea09.png">

<img width="300" src="https://user-images.githubusercontent.com/22608780/62442523-cffc8400-b775-11e9-9135-82f6efcb882e.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/62442533-d1c64780-b775-11e9-9a24-f273ef4dd798.png">

Update release notes:

- [x] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

anitaa1990 committed:
Added date bar UI to old stats view

anitaa1990 committed:
Fix MyStorePresenterTest class unit test imports

anitaa1990 committed:
Added method to DateUtils to get a display date in the format 'MMM yyyy'

anitaa1990 committed:
Added logic to display/update/clear date range for each tab

anitaa1990 committed:
Added method to DateUtils to format dates based on granularity

anitaa1990 committed:
Added logic to display/update/clear date range for each tab in v4 stats UI

anitaa1990 committed:
Added release notes

anitaa1990 committed:
Fix lint errors

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/util/DateUtils.kt:
```diff
@@ -14,6 +14,7 @@ import java.util.Locale
 
 object DateUtils {
     val friendlyMonthDayFormat by lazy { SimpleDateFormat("MMM d", Locale.getDefault()) }
+    val friendlyDayMonthDateFormat by lazy { SimpleDateFormat("EEEE, MMM d", Locale.ROOT) }
```
Just curious why we use `Locale.ROOT` here when we use `Local.getDefault()` elsewhere?

anitaa1990 replied:
I keep getting this lint warning when using `Locale.getDefault()`. So I had modified it `Locale.ROOT`

<img src="https://user-images.githubusercontent.com/22608780/62456369-e3b7e280-b795-11e9-9833-fc5091b05b75.png">

nbradbury replied:
That warning is valid, but does using `ROOT` really resolve it? In either case, I think we want to be consistent here.

anitaa1990 replied:
You're right! It _is_ actually recommended to use `Locale.getDefault()` in most cases so I modified the logic in fd4f03a

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/util/DateUtils.kt:
```diff
@@ -229,4 +249,55 @@ object DateUtils {
+    @Throws(IllegalArgumentException::class)
+    fun getMonthString(iso8601Date: String): String {
+        return try {
+            val (_, month, _) = iso8601Date.split("-")
```
Until I saw this I had no idea Kotlin allowed syntax like `val (_, month, _)`. 👍

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/dashboard/DashboardStatsView.kt:
```diff
@@ -320,6 +326,26 @@ class DashboardStatsView @JvmOverloads constructor(ctx: Context, attrs: Attribut
+     * If the start and end date are the same i.e. 2019 for YEARS, then only display
+     * the date and not the range
+     */
+    private fun updateDateRangeView() {
```
This seems to somewhat duplicate the logic [here](https://github.com/woocommerce/woocommerce-android/blob/3fa5718e0e7887e1a2e9ec7ae3c5a919f33e424c/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreDateRangeView.kt#L23). Do we need both of these functions or can they be consolidated?

anitaa1990 replied:
I had first implemented a util class that can be reused by both UI's. But the logic on formatting the date string based on the granularity, is pretty different between both the UIs. And also the input params for both the methods are different. For instance, the new v4 stats UI, would require the input parameter `WCRevenueStatsModel` whereas the old stats using a map. So I thought it would be better to keep them separate.

nbradbury commented on WooCommerce/src/main/kotlin/com/woocommerce/android/ui/mystore/MyStoreDateRangeView.kt:
```diff
@@ -0,0 +1,58 @@
+            StatsGranularity.WEEKS -> {
+                val endInterval = revenueStatsModel?.getIntervalList()?.last()?.interval
+                val endDate = endInterval?.let { getDateValue(it, granularity) }
+                String.format("%s–%s", startDate, endDate)
```
np: I think it looks better to put spaces around the hyphen.

![Screenshot_1564996937](https://user-images.githubusercontent.com/3903757/62453825-27343180-b741-11e9-897b-082cc924b45f.png)

anitaa1990 replied:
👍 I fixed this in  dc45995 and faaa2fb

anitaa1990 committed:
Add space between the hyphen when displaying the date range

anitaa1990 committed:
Add space between the hyphen when displaying the date range in old UI

anitaa1990 commented:
Thank you for the quick review @nbradbury!! I've fixed the nit picks and addressed the comments. PR ready for another round 👍

anitaa1990 committed:
Added the recommended Locale.getDefault() when formatting date

nbradbury commented:
Looks good! I'll merge once this has passed design review, and I have a related question for @kyleaparker: now that we show the date range at the top is it still necessary to show the dates at the bottom?

![Screenshot_1565088493](https://user-images.githubusercontent.com/3903757/62534223-992a7a80-b816-11e9-9746-661513d6daa4.png)

anitaa1990 commented:
Thanks for the quick review @nbradbury ! 

> now that we show the date range at the top is it still necessary to show the dates at the bottom?

I completely defer to @kyleaparker on this but looking at the designs for this, I believe we need to modify the x-axis and labels for each tab. 
<img width="300" src="https://user-images.githubusercontent.com/22608780/62538195-d4a56300-b86f-11e9-9c5b-d8745b2e2701.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/62538210-df5ff800-b86f-11e9-95b6-2a6b93e4012f.png">

<img width="300" src="https://user-images.githubusercontent.com/22608780/62538229-e981f680-b86f-11e9-896a-0e8a0fbcdd00.png"> .  <img width="300"  src="https://user-images.githubusercontent.com/22608780/62538250-f3a3f500-b86f-11e9-855e-db0eb089f55e.png">

There is a separate issue [here](https://github.com/woocommerce/woocommerce-android/issues/1260), to modify the x-axis of the charts for each tab.

nbradbury commented:
> There is a separate issue [here](https://github.com/woocommerce/woocommerce-android/issues/1260), to modify the x-axis of the charts for each tab.

Ah, I had completely missed that! @kyleaparker Ignore my question above - and please ping me once this passes design review so I can merge it.

kyleaparker commented:
Looks good! Just one question, is the date font using medium or bold?

nbradbury commented:
> Is the date font using medium or bold

@kyleaparker It's bold.

kyleaparker commented:
Ok that will just need to be changed to Medium :)

nbradbury commented:
> Ok that will just need to be changed to Medium :)

Okay, here's how that looks - good?

![normal](https://user-images.githubusercontent.com/3903757/62574695-11b92780-b867-11e9-83cb-9a65e3f41757.png)

kyleaparker commented:
That looks like regular to me, is there an in-between setting? I'm using medium in the designs so just needs to match that

nbradbury commented:
> That looks like regular to me, is there an in-between setting? I'm using medium in the designs so just needs to match that

Oh, sorry - I misunderstood what you meant. Here's how Medium looks:

![medium](https://user-images.githubusercontent.com/3903757/62576128-3c58af80-b86a-11e9-9579-70835b0085f1.png)

kyleaparker commented:
Looks good! Sorry one other comment, the alignment to the left of the date looks slightly off, what is the margin set at?

nbradbury commented:
> the alignment to the left of the date looks slightly off, what is the margin set at?

It's set to `default_padding`, which is `8dp` on phones and `16dp` on tablets.

kyleaparker commented:
It should align with the other text: 

<img width="539" alt="62576128-3c58af80-b86a-11e9-9579-70835b0085f1 png (540×960) 2019-08-06 17-06-32" src="https://user-images.githubusercontent.com/712873/62577280-9bb7bf00-b86c-11e9-8277-f1adb92021e7.png">

nbradbury commented:
> It should align with the other text:

Okay, I'm just shutting down for the day so I'll leave this for @anitaa1990 to finish. Sorry @anitaa1990 - I was hoping to get this merged today!

anitaa1990 committed:
Modified the styles for the date range bar

anitaa1990 commented:
No worries @nbradbury! Thank you!!

I modified the padding for the date range bar and removed the `bold` settings from the text. This is how it looks now. I'm not sure what I need to do to make text `medium` though :) @nbradbury could I trouble you to let me know what you changed for this?

<img width="300" src="https://user-images.githubusercontent.com/22608780/62609311-23f69c80-b91f-11e9-98c6-bc01b56b4f88.png">

nbradbury commented:
> @nbradbury could I trouble you to let me know what you changed for this?

Sorry, I should've checked that in! Just add this to the style:

```
<item name="android:fontFamily">@font/roboto_medium</item>
```

anitaa1990 committed:
Added roboto medium font for the date range bar

anitaa1990 commented:
Thank you for the tip @nbradbury! I totally misunderstood that! I've added the font in 0dcb6a3.

@kyleaparker this is how the screen looks now:
(LEFT: current stats . RIGHT: new stats)
<img width="300" src="https://user-images.githubusercontent.com/22608780/62620404-e0a82800-b936-11e9-9a5d-5cc93eefada7.png"> . <img width="300" src="https://user-images.githubusercontent.com/22608780/62620407-e271eb80-b936-11e9-91ff-03ddfb8ee8ab.png">

anitaa1990 commented:
Thank you @kyleaparker!! @nbradbury, I see a thumbs up from Kylea. PR ready for another round 👍

nbradbury commented:
:shipit: !
